libsolidity/interface/CompilerStack.cpp

```diff
@@ -331,6 +331,16 @@ bool CompilerStack::analyze()
 		if (source->ast)
 			Scoper::assignScopes(*source->ast);
 
+	bool ok = analyzeInternal();
+
+	m_stackState = AnalysisPerformed;
```

Comment:
@chriseth Can you check this line? This leads to some changes in tests https://github.com/ethereum/solidity/pull/9959/files#r528634568

Reply:
Previously, before daba02b,  in `CompilerStack::analyzeInternal()`, in case of several early `return false;` we did not set the state `m_stackState = AnalysisPerformed`

For instance here, for a *non-fatal error* we just returned `false`:

```
bool CompilerStack::analyzeInternal()
{
    . . . .
        for (Source const* source: m_sourceOrder)
            if (source->ast && !resolver.resolveNamesAndTypes(*source->ast))
                return false;
```

It is somewhat strange, because a *fatal error* in `resolveNamesAndTypes`
would be caught and the state `m_stackState = AnalysisPerformed` would be set:

```
    catch (FatalError const&)
    {
        if (m_errorReporter.errors().empty())
            throw; // Something is weird here, rather throw again.
        noErrors = false;
    }

    m_stackState = AnalysisPerformed;
    if (!noErrors)
        m_hasError = true;

    return !m_hasError;
}
```

I think it has to be consistent, and we should
- either set `m_stackState = AnalysisPerformed` when `CompilerStack::analyze()` *finishes, always*;
- or do it if `CompilerStack::analyze()` *finishes successfully*.

Reply:
I think the old behaviour was by design and is inherited from the efforts to error-recovery...
``AnalysisPerformed`` is not meant to imply "Analysis successful"... that's why there's ``m_hasError`` checks in cricital stages that want to move beyond ``AnalysisPerformed``. Not sure how sane any of that really is, though...

But I guess this change is independent of fixing the ICE by moving the check? Maybe we should split that...